File: src/main/java/org/saigon/striker/controller/UserController.java
package org.saigon.striker.controller;

import org.saigon.striker.model.User;
import org.saigon.striker.service.UserService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.util.UriTemplate;
import reactor.core.publisher.Mono;

import javax.validation.Valid;

import static org.apache.commons.lang3.Validate.notNull;
import static org.saigon.striker.model.UserModelsKt.toEntity;
import static org.saigon.striker.model.UserModelsKt.toUser;

@RestController
public class UserController {

    public static final String USERS_URI_TEMPLATE = "/admin/user";
    public static final String USER_URI_TEMPLATE = USERS_URI_TEMPLATE + "/{userId}";

    private final UserService userService;

    public UserController(UserService userService) {
        this.userService = notNull(userService);
    }

    @PostMapping(USERS_URI_TEMPLATE)
    public Mono<ResponseEntity<User>> createUser(@Valid @RequestBody User user) {
        return userService.createUser(toEntity(user))
                .map(userEntity -> ResponseEntity
                        .created(new UriTemplate(USER_URI_TEMPLATE).expand(userEntity.getId()))
                        .body(toUser(userEntity).withoutPassword()));
    }

    @GetMapping(USER_URI_TEMPLATE)
    public Mono<ResponseEntity<User>> getUser(@PathVariable String userId) {
        return userService.getUser(userId)
                .map(userEntity -> ResponseEntity.ok(toUser(userEntity).withoutPassword()))
                .defaultIfEmpty(ResponseEntity.notFound().build());
    }

    @DeleteMapping(USER_URI_TEMPLATE)
    public Mono<ResponseEntity<?>> deleteUser(@PathVariable String userId) {
        return userService.deleteUser(userId)
                .thenReturn(ResponseEntity.noContent().build());
    }
}

File: src/main/kotlin/org/saigon/striker/controller/UserController.kt
package org.saigon.striker.controller

import org.saigon.striker.model.User
import org.saigon.striker.model.toEntity
import org.saigon.striker.model.toUserWithoutPassword
import org.saigon.striker.service.UserService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
class UserController(val userService: UserService) {

    @PostMapping(UriTemplates.USERS)
    suspend fun createUser(@RequestBody user: User): ResponseEntity<User> {
        val userEntity = userService.createUser(user.toEntity())
        return ResponseEntity.created(UriTemplates.expandUser(userEntity.id))
            .body(userEntity.toUserWithoutPassword())
    }

    @GetMapping(UriTemplates.USER)
    suspend fun getUser(@PathVariable userId: String): ResponseEntity<User> {
        val userEntity = userService.getUser(userId) ?: return ResponseEntity.notFound().build()
        return ResponseEntity.ok(userEntity.toUserWithoutPassword())
    }

    @DeleteMapping(UriTemplates.USER)
    suspend fun deleteUser(@PathVariable userId: String): ResponseEntity<Any> {
        userService.deleteUser(userId)
        return ResponseEntity.noContent().build()
    }
}
